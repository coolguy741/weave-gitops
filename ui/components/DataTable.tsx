import {
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
} from "@material-ui/core";
import _ from "lodash";
import * as React from "react";
import styled from "styled-components";
import Text from "./Text";

/** DataTable Properties  */
export interface Props {
  /** CSS MUI Overrides or other styling. */
  className?: string;
  /** A list of objects with two fields: `label`, which is a string representing the column header, and `value`, which can be a string, or a function that extracts the data needed to fill the table cell. */
  fields: {
    label: string;
    value: string | ((k: any) => string | JSX.Element);
  }[];
  /** A list of data that will be iterated through to create the columns described in `fields`. */
  rows: any[];
  /** A list of strings representing the sortable columns of the table, passed into lodash's `_.sortBy`. */
  sortFields: string[];
  /** Indicates whether to reverse the sorted array. */
  reverseSort?: boolean;
}

const EmptyRow = styled(TableRow)<{ colSpan: number }>`
  font-style: italic;

  td {
    text-align: center;
  }
`;

/** Form DataTable */
function UnstyledDataTable({
  className,
  fields,
  rows,
  sortFields,
  reverseSort,
}: Props) {
  const sorted = _.sortBy(rows, sortFields);

  if (reverseSort) {
    sorted.reverse();
  }

  const r = _.map(sorted, (r, i) => (
    <TableRow key={i}>
      {_.map(fields, (f) => (
        <TableCell key={f.label}>
          <Text>{typeof f.value === "function" ? f.value(r) : r[f.value]}</Text>
        </TableCell>
      ))}
    </TableRow>
  ));

  return (
    <div className={className}>
      <TableContainer>
        <Table aria-label="simple table">
          <TableHead>
            <TableRow>
              {_.map(fields, (f) => (
                <TableCell key={f.label}>{f.label}</TableCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {r.length > 0 ? (
              r
            ) : (
              <EmptyRow colSpan={fields.length}>
                <TableCell colSpan={fields.length}>
                  <span style={{ fontStyle: "italic" }}>No rows</span>
                </TableCell>
              </EmptyRow>
            )}
          </TableBody>
        </Table>
      </TableContainer>
    </div>
  );
}

export const DataTable = styled(UnstyledDataTable)`
  th {
    font-weight: bold;
  }
`;

export default DataTable;
