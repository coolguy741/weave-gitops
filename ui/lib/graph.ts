// A collection of helper functions to render a graph of kubernetes objects
// with in the context of their parent-child relationships.
import {
  Applications,
  GroupVersionKind,
  UnstructuredObject,
} from "./api/applications/applications.pb";

export type UnstructuredObjectWithParent = UnstructuredObject & {
  parentUid?: string;
};

// Kubernetes does not allow us to query children by parents.
// We keep a list of common parent-child relationships
// to look up children recursively.
export const PARENT_CHILD_LOOKUP = {
  Deployment: {
    group: "apps",
    version: "v1",
    kind: "Deployment",
    children: [
      {
        group: "apps",
        version: "v1",
        kind: "ReplicaSet",
        children: [{ version: "v1", kind: "Pod" }],
      },
    ],
  },
};

export const getChildrenRecursive = async (
  appsClient: typeof Applications,
  result: UnstructuredObjectWithParent[],
  object: UnstructuredObjectWithParent,
  lookup: any
) => {
  result.push(object);

  const k = lookup[object.groupVersionKind.kind];

  if (k && k.children) {
    for (let i = 0; i < k.children.length; i++) {
      const child: GroupVersionKind = k.children[i];

      const res = await appsClient.GetChildObjects({
        parentUid: object.uid,
        groupVersionKind: child,
      });

      for (let q = 0; q < res.objects.length; q++) {
        const c = res.objects[q];

        // Dive down one level and update the lookup accordingly.
        await getChildrenRecursive(
          appsClient,
          result,
          { ...c, parentUid: object.uid },
          {
            [child.kind]: child,
          }
        );
      }
    }
  }
};
