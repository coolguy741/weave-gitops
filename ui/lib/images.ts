/* eslint-disable */
// @ts-nocheck

import failedSrc from "data-url:../images/failed.svg";
import fancyCardBackgroundBlue from "data-url:../images/fancy-card-background-blue.svg";
import fancyCardBackgroundOrange from "data-url:../images/fancy-card-background-orange.svg";
import fancyCardBackground from "data-url:../images/fancy-card-background.svg";
import bg from "data-url:../images/General-backgroundok.png";
import logoSrc from "data-url:../images/logo.svg";
import signInWheel from "data-url:../images/SignInWheel.svg";
import successSrc from "data-url:../images/success.svg";
import suspendedSrc from "data-url:../images/suspended.svg";
import titleSrc from "data-url:../images/Title.svg";
import weaveLogo from "data-url:../images/WeaveLogo.svg";

export default {
  bg,
  failedSrc,
  fancyCardBackground,
  fancyCardBackgroundBlue,
  fancyCardBackgroundOrange,
  logoSrc,
  signInWheel,
  successSrc,
  suspendedSrc,
  titleSrc,
  weaveLogo,
};
