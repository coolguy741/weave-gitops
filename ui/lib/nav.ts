import qs from "query-string";
import { SourceRefSourceKind } from "./api/core/types.pb";
import { NoNamespace, PageRoute, V2Routes } from "./types";

// getParentNavValue returns the parent for a child page.
// This keeps the nav element highlighted if we are on a child page.
// Example: /sources and /git_repo will both show the "Sources" nav as selected.
export const getParentNavValue = (
  currentPage: any
): PageRoute | V2Routes | boolean => {
  switch (currentPage) {
    case V2Routes.Automations:
    case V2Routes.Kustomization:
    case V2Routes.HelmRelease:
      return V2Routes.Automations;

    case V2Routes.Sources:
    case V2Routes.GitRepo:
    case V2Routes.HelmChart:
    case V2Routes.HelmRepo:
    case V2Routes.Bucket:
      return V2Routes.Sources;

    case V2Routes.FluxRuntime:
      return V2Routes.FluxRuntime;

    default:
      // The "Tabs" component of material-ui wants a bool
      return false;
  }
};

const pageTitles = {
  [V2Routes.Automations]: "Applications",
  [V2Routes.Sources]: "Sources",
  [V2Routes.FluxRuntime]: "Flux Runtime",
};

export const getPageLabel = (route: V2Routes): string => {
  return pageTitles[route];
};

export const formatURL = (page: string, query: any = {}) => {
  return `${page}?${qs.stringify(query)}`;
};

export const formatSourceURL = (
  kind: SourceRefSourceKind,
  name: string,
  namespace: string = NoNamespace
) => {
  return formatURL(sourceTypeToRoute(kind), { name, namespace });
};

export function sourceTypeToRoute(t: SourceRefSourceKind): V2Routes {
  switch (t) {
    case SourceRefSourceKind.GitRepository:
      return V2Routes.GitRepo;

    case SourceRefSourceKind.Bucket:
      return V2Routes.Bucket;

    case SourceRefSourceKind.HelmRepository:
      return V2Routes.HelmRepo;

    case SourceRefSourceKind.HelmChart:
      return V2Routes.HelmChart;

    default:
      break;
  }
}
