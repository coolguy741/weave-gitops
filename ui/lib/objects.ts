import { stringify } from "yaml";
import { FluxObjectKind, Object as ResponseObject } from "./api/core/types.pb";

export enum Kind {
  GitRepository = "GitRepository",
  Bucket = "Bucket",
  HelmRepository = "HelmRepository",
  HelmChart = "HelmChart",
  Kustomization = "Kustomization",
  HelmRelease = "HelmRelease",
}

export function fluxObjectKindToKind(fok: FluxObjectKind): Kind {
  return Kind[FluxObjectKind[fok].slice(4)];
}

export class FluxObject {
  obj: any;

  constructor(response: ResponseObject) {
    this.obj = JSON.parse(response.payload);
  }

  yaml() {
    return stringify(this.obj);
  }

  // Return list of key-value pairs for the metadata annotations that follow
  // our spec
  metadata(): [string, string][] {
    const prefix = "metadata.weave.works/";
    const annotations = this.obj.metadata.annotations || {};
    return Object.keys(annotations).flatMap((key) => {
      if (!key.startsWith(prefix)) {
        return [];
      } else {
        return [[key.slice(prefix.length), annotations[key] as string]];
      }
    });
  }
}
