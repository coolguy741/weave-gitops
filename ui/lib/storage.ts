import { GitProvider } from "./api/applications/applications.pb";
import { PageRoute } from "./types";

const tokenKey = (providerName: GitProvider) =>
  `gitProviderToken_${providerName}`;

export function storeProviderToken(providerName: GitProvider, token: string) {
  if (!window.localStorage) {
    console.warn("no local storage found");
    return;
  }

  localStorage.setItem(tokenKey(providerName), token);
}

export function getProviderToken(providerName: GitProvider): string {
  if (!window.localStorage) {
    console.warn("no local storage found");
    return;
  }

  return localStorage.getItem(tokenKey(providerName));
}

const CALLBACK_STATE_KEY = "oauth_callback_state";

export type CallbackSessionState = { page: PageRoute; state: any };

export function storeCallbackState(value: CallbackSessionState) {
  if (!window.sessionStorage) {
    console.warn("no session storage found");
    return;
  }

  if (!value) {
    return null;
  }

  return sessionStorage.setItem(CALLBACK_STATE_KEY, JSON.stringify(value));
}

export function getCallbackState(): CallbackSessionState {
  const state = sessionStorage.getItem(CALLBACK_STATE_KEY);

  if (!state) {
    return null;
  }

  try {
    return JSON.parse(state);
  } catch (e) {
    return null;
  }
}

export function clearCallbackState() {
  return sessionStorage.removeItem(CALLBACK_STATE_KEY);
}
