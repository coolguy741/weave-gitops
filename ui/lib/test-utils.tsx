import { MuiThemeProvider } from "@material-ui/core";
import { createMemoryHistory } from "history";
import _ from "lodash";
import * as React from "react";
import { QueryClient, QueryClientProvider } from "react-query";
import { Router } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import AppContextProvider, { AppProps } from "../contexts/AppContext";
import { CoreClientContext } from "../contexts/CoreClientContext";
import {
  Applications,
  GetGithubAuthStatusRequest,
  GetGithubAuthStatusResponse,
  GetGithubDeviceCodeRequest,
  GetGithubDeviceCodeResponse,
  ParseRepoURLRequest,
  ParseRepoURLResponse,
  ValidateProviderTokenRequest,
  ValidateProviderTokenResponse,
} from "./api/applications/applications.pb";
import {
  Core,
  GetChildObjectsRequest,
  GetChildObjectsResponse,
  GetReconciledObjectsRequest,
  GetReconciledObjectsResponse,
  GetVersionRequest,
  GetVersionResponse,
  ListObjectsRequest,
  ListObjectsResponse,
} from "./api/core/core.pb";
import theme, { muiTheme } from "./theme";
import { RequestError } from "./types";

export type CoreOverrides = {
  GetChildObjects?: (req: GetChildObjectsRequest) => GetChildObjectsResponse;
  GetReconciledObjects?: (
    req: GetReconciledObjectsRequest
  ) => GetReconciledObjectsResponse;
  GetVersion?: (req: GetVersionRequest) => GetVersionResponse;
  ListObjects?: (req: ListObjectsRequest) => ListObjectsResponse;
};

export const createCoreMockClient = (
  ovr: CoreOverrides,
  error?: RequestError
): typeof Core => {
  const promisified = _.reduce(
    ovr,
    (result, handlerFn, method) => {
      result[method] = (req) => {
        if (error) {
          return new Promise((_, reject) => reject(error));
        }
        return new Promise((accept) => accept(handlerFn(req) as any));
      };

      return result;
    },
    {}
  );

  return promisified as typeof Core;
};

export type ApplicationOverrides = {
  GetGithubDeviceCode?: (
    req: GetGithubDeviceCodeRequest
  ) => GetGithubDeviceCodeResponse;
  GetGithubAuthStatus?: (
    req: GetGithubAuthStatusRequest
  ) => GetGithubAuthStatusResponse;
  ParseRepoURL?: (req: ParseRepoURLRequest) => ParseRepoURLResponse;
  ValidateProviderToken?: (
    req: ValidateProviderTokenRequest
  ) => ValidateProviderTokenResponse;
};

// Don't make the user wire up all the promise stuff to be interface-compliant
export const createMockClient = (
  ovr: ApplicationOverrides,
  error?: RequestError
): typeof Applications => {
  const promisified = _.reduce(
    ovr,
    (result, handlerFn, method) => {
      result[method] = (req) => {
        if (error) {
          return new Promise((_, reject) => reject(error));
        }
        return new Promise((accept) => accept(handlerFn(req) as any));
      };

      return result;
    },
    {}
  );

  return promisified as typeof Applications;
};

export function withTheme(element) {
  return (
    <MuiThemeProvider theme={muiTheme}>
      <ThemeProvider theme={theme}>{element}</ThemeProvider>
    </MuiThemeProvider>
  );
}

type TestContextProps = AppProps & { api?: typeof Core };

export function withContext(
  TestComponent,
  url: string,
  { api, ...appProps }: TestContextProps
) {
  const history = createMemoryHistory();
  history.push(url);
  const queryClient = new QueryClient({
    defaultOptions: { queries: { retry: false } },
  });
  const isElement = React.isValidElement(TestComponent);
  return (
    <Router history={history}>
      <AppContextProvider renderFooter {...appProps}>
        <QueryClientProvider client={queryClient}>
          <CoreClientContext.Provider value={{ api, featureFlags: {} }}>
            {isElement ? TestComponent : <TestComponent />}
          </CoreClientContext.Provider>
        </QueryClientProvider>
      </AppContextProvider>
    </Router>
  );
}
