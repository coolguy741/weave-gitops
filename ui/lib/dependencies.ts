import { Automation } from "../hooks/automations";
import { FluxObjectNode } from "./objects";

// findNode searches for a node by the node's name and namespace.
export function findNode(
  nodes: FluxObjectNode[],
  name: string,
  namespace: string
): FluxObjectNode | null {
  const matches = nodes.filter(
    (node) => node.name === name && node.namespace === namespace
  );

  if (matches.length > 0) {
    return matches[0];
  } else {
    return null;
  }
}

// getNeighborNodes returns nodes which depend on the current node
// or are dependencies of the current node.
export function getNeighborNodes(
  nodes: FluxObjectNode[],
  currentNode: FluxObjectNode
): FluxObjectNode[] {
  const dependencyNodes = currentNode.dependsOn
    .map((dependency) => {
      const name = dependency.name;
      let namespace = dependency.namespace;

      if (!namespace) {
        namespace = currentNode.namespace;
      }

      return findNode(nodes, name, namespace);
    })
    .filter((n) => n);

  const dependentNodes = nodes.filter((node) => {
    let isDependent = false;

    for (const dependency of node.dependsOn) {
      const name = dependency.name;
      let namespace = dependency.namespace;
      if (!namespace) {
        namespace = node.namespace;
      }

      if (name === currentNode.name && namespace === currentNode.namespace) {
        isDependent = true;
        break;
      }
    }

    return isDependent;
  });

  return dependencyNodes.concat(dependentNodes);
}

// getGraphNodes returns all nodes in the current node's dependency tree, including the current node.
export function getGraphNodes(
  nodes: FluxObjectNode[],
  automation: Automation
): FluxObjectNode[] {
  // Find node, corresponding to the automation.
  const currentNode = findNode(nodes, automation.name, automation.namespace);

  if (!currentNode) {
    return [];
  }

  currentNode.isCurrentNode = true;

  // Find nodes in the current node's dependency tree.
  let graphNodes: FluxObjectNode[] = [];

  const visitedNodes = new Map<string, boolean>();

  let nodesToExplore: FluxObjectNode[] = [currentNode].concat(
    getNeighborNodes(nodes, currentNode)
  );

  if (nodesToExplore.length === 1) {
    return [];
  }

  while (nodesToExplore.length > 0) {
    const node = nodesToExplore.shift();

    const newNodes = getNeighborNodes(nodes, node).filter(
      (n) => !visitedNodes[n.id]
    );

    for (const n of newNodes) {
      visitedNodes[n.id] = true;
    }

    nodesToExplore = nodesToExplore.concat(newNodes);

    graphNodes = graphNodes.concat(node);
  }

  // Remove duplicates from graphNodes.
  graphNodes = graphNodes.filter(
    (node, index) => graphNodes.indexOf(node) === index
  );

  return graphNodes;
}
