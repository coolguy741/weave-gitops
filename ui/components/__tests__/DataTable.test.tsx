import { fireEvent, render, screen } from "@testing-library/react";
import "jest-styled-components";
import React from "react";
import renderer from "react-test-renderer";
import { withTheme } from "../../lib/test-utils";
import DataTable from "../DataTable";

describe("DataTable", () => {
  const rows = [
    {
      name: "the-cool-app",
      status: "Ready",
      lastUpdate: "2006-01-02T15:04:05-0700",
    },
    {
      name: "podinfo",
      status: "Failed",
      lastUpdate: "2006-01-02T15:04:05-0700",
    },
    {
      name: "nginx",
      status: "Ready",
      lastUpdate: "2006-01-02T15:04:05-0700",
    },
  ];

  const fields = [
    {
      label: "Name",
      value: ({ name }) => <a href="/some_url">{name}</a>,
    },
    {
      label: "Status",
      value: (v) => v.status,
    },
    {
      label: "Last Updated",
      value: "lastUpdate",
    },
  ];
  describe("sorting", () => {
    it("initially sorts based on sortFields[0]", () => {
      render(
        withTheme(
          <DataTable
            sortFields={["name", "status"]}
            fields={fields}
            rows={rows}
          />
        )
      );
      const firstRow = screen.getAllByRole("row")[1];
      expect(firstRow.innerHTML).toMatch(/nginx/);
    });
    it("reverses sort on thead click", () => {
      render(
        withTheme(
          <DataTable
            sortFields={["name", "status"]}
            fields={fields}
            rows={rows}
          />
        )
      );

      const nameButton = screen.getByText("Name");
      fireEvent.click(nameButton);
      const firstRow = screen.getAllByRole("row")[1];
      expect(firstRow.innerHTML).toMatch(/the-cool-app/);
    });
    it("resets reverseSort and switches sort column on different thead click", () => {
      render(
        withTheme(
          <DataTable
            sortFields={["name", "status"]}
            fields={fields}
            rows={rows}
          />
        )
      );
      const nameButton = screen.getByText("Name");
      fireEvent.click(nameButton);
      const statusButton = screen.getByText("Status");
      fireEvent.click(statusButton);
      const firstRow = screen.getAllByRole("row")[1];
      expect(firstRow.innerHTML).toMatch(/podinfo/);
    });
  });
  describe("snapshots", () => {
    it("renders", () => {
      const tree = renderer
        .create(
          withTheme(
            <DataTable sortFields={["name"]} fields={fields} rows={rows} />
          )
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
