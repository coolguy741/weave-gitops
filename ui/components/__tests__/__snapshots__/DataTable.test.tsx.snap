// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable snapshots renders 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  height: 100%;
  width: 100%;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  height: 100%;
  width: 100%;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c11 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  text-transform: none;
  font-style: normal;
}

.c16 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 20px;
  font-weight: 400;
  text-transform: none;
  font-style: normal;
  color: #737373;
}

.c9 svg {
  fill: !important;
  height: 16px;
  width: 16px;
}

.c9.downward {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c9.upward {
  -webkit-transform: initial;
  -ms-transform: initial;
  transform: initial;
}

.c9 .c10 {
  margin-left: 4px;
}

.c5.MuiButton-root {
  height: 32px;
  font-size: 12px;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  line-height: 1;
  border-radius: 2px;
  font-weight: 600;
}

.c5.MuiButton-outlined {
  padding: 8px 12px;
  border-color: #d8d8d8;
}

.c7 {
  padding: 4px;
}

.c12 {
  position: relative;
  height: 100%;
  width: 0px;
  left: 24px;
  -webkit-transition-property: width,left;
  transition-property: width,left;
  -webkit-transition-duration: 0.5s;
  transition-duration: 0.5s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c12.open {
  left: 0px;
  width: 350px;
}

.c13 {
  background: rbga(255,255,255,0.75);
  height: 100%;
  width: 100%;
  border-left: 2px solid #d8d8d8;
  padding: 24px;
  padding-left: 32px;
}

.c15 .MuiListItem-gutters {
  padding-left: 0px;
}

.c15 .MuiCheckbox-root {
  padding: 0px;
}

.c15 .MuiCheckbox-colorSecondary.Mui-checked {
  color: #00b3ec;
}

.c6.MuiButton-root {
  margin: 0;
  text-transform: none;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
}

.c6.MuiButton-text {
  min-width: 0px;
}

.c6.MuiButton-text .selected {
  color: #1a1a1a;
}

.c6.arrow {
  min-width: 0px;
}

.c1 {
  width: 100%;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  overflow-x: hidden;
}

.c1 h2 {
  padding: 8px;
  font-size: 14px;
  font-weight: 600;
  color: #737373;
  margin: 0px;
  white-space: nowrap;
}

.c1 .MuiTableRow-root {
  -webkit-transition: background 0.5s ease-in-out;
  transition: background 0.5s ease-in-out;
}

.c1 .MuiTableRow-root:not(.MuiTableRow-head):hover {
  background: #f5f5f5;
  -webkit-transition: background 0.5s ease-in-out;
  transition: background 0.5s ease-in-out;
}

.c1 table {
  margin-top: 12px;
}

.c1 th {
  padding: 8px;
  background: #F6F7F9;
}

.c1 td {
  padding-left: 24px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c1 .filter-options-chip {
  background-color: #E5F7FD;
}

<div
  className="c0 c1"
>
  <div
    className="c2"
  />
  <div
    className="c3"
  >
    <div
      className="MuiTableContainer-root"
    >
      <table
        aria-label="simple table"
        className="MuiTable-root"
        role={null}
      >
        <thead
          className="MuiTableHead-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root MuiTableRow-head"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head"
              scope="col"
            >
              <div
                className="c4"
              >
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-text c5 c6 MuiButton-colorInherit MuiButton-disableElevation"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <span
                    className="MuiButton-label"
                  >
                    <h2
                      className="selected"
                    >
                      Name
                    </h2>
                  </span>
                </button>
                <div
                  className="c7"
                />
                <div
                  className="c8 c9 downward"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                    />
                  </svg>
                </div>
              </div>
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head"
              scope="col"
            >
              <div
                className="c4"
              >
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-text c5 c6 MuiButton-colorInherit MuiButton-disableElevation"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <span
                    className="MuiButton-label"
                  >
                    <h2
                      className=""
                    >
                      Status
                    </h2>
                  </span>
                </button>
                <div
                  className="c7"
                />
                <div
                  style={
                    Object {
                      "width": "16px",
                    }
                  }
                />
              </div>
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head"
              scope="col"
            >
              <div
                className="c4"
              >
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-text c5 c6 MuiButton-colorInherit MuiButton-disableElevation"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <span
                    className="MuiButton-label"
                  >
                    <h2
                      className=""
                    >
                      Last Updated
                    </h2>
                  </span>
                </button>
                <div
                  className="c7"
                />
                <div
                  style={
                    Object {
                      "width": "16px",
                    }
                  }
                />
              </div>
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head"
              scope="col"
            >
              <div
                className="c4"
              >
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-text c5 c6 MuiButton-colorInherit MuiButton-disableElevation"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <span
                    className="MuiButton-label"
                  >
                    <h2
                      className=""
                    >
                      Last Synced At
                    </h2>
                  </span>
                </button>
                <div
                  className="c7"
                />
                <div
                  style={
                    Object {
                      "width": "16px",
                    }
                  }
                />
              </div>
            </th>
          </tr>
        </thead>
        <tbody
          className="MuiTableBody-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              <span
                className="c10 c11"
                size="medium"
              >
                <a
                  href="/some_url"
                >
                  nginx
                </a>
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              <span
                className="c10 c11"
                size="medium"
              >
                Ready
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              <span
                className="c10 c11"
                size="medium"
              >
                2004-01-02T15:04:05-0700
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              <span
                className="c10 c11"
                size="medium"
              >
                3000
              </span>
            </td>
          </tr>
          <tr
            className="MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              <span
                className="c10 c11"
                size="medium"
              >
                <a
                  href="/some_url"
                >
                  podinfo
                </a>
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              <span
                className="c10 c11"
                size="medium"
              >
                -
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              <span
                className="c10 c11"
                size="medium"
              >
                2006-01-02T15:04:05-0700
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              <span
                className="c10 c11"
                size="medium"
              >
                2000
              </span>
            </td>
          </tr>
          <tr
            className="MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              <span
                className="c10 c11"
                size="medium"
              >
                <a
                  href="/some_url"
                >
                  the-cool-app
                </a>
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              <span
                className="c10 c11"
                size="medium"
              />
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              <span
                className="c10 c11"
                size="medium"
              >
                2005-01-02T15:04:05-0700
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              <span
                className="c10 c11"
                size="medium"
              >
                1000
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      className="c12"
      data-testid="container"
    >
      <div
        className="c13"
      >
        <div
          className="c14 c15"
        >
          <div
            className="c2"
          >
            <span
              className="c10 c16"
              color="neutral30"
              size="large"
            >
              Filters
            </span>
          </div>
          <form
            className="ControlledForm"
            onSubmit={[Function]}
          >
            <ul
              className="MuiList-root MuiList-padding"
            />
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
